### DISCUSSION TITLE
ANALOG#A0div10 rule not firing for analog photoresistor

### DISCUSSION THREAD

**Discussion by aheath70** [arendst/Tasmota#17131] [Q&A] [answered]
### PROBLEM DESCRIPTION
I have a generic ESP8266 with a KY-018 photoresistor (https://arduinomodules.info/ky-018-photoresistor-module/) attached to A0, 3.3v and ground. It is configured in Tasmota as ADC Light. It is successfully reporting illuminance with almost instantaneous updates in the Tasmota web interface, but there is a delay of up 5 minutes in sending an MQTT message line with my TelePeriod. I have a use case where I want to trigger the MQTT message as soon as the illuminance changes. 

I have attempted to implement the below rule to force telemetry to send via MQTT but it is not being fired. I have covered and uncovered the sensor causing significant change to the illumination reported in the web interface but I cannot get the rule to fire.

```
rule1 ON ANALOG#A0div10 DO status 10 ENDON
rule1 1
```

I have tested other rules such as when a relay is powered on sending a MQTT message to a specific topic, and that rule works, so rules are enabled. I am using the standard Tasmota image from the web installer on the Tasmota web site.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Nodemcu V2 generic ESP8266
- [x] Tasmota binary firmware version number used: 12.2.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: Tasmota web flasher at https://tasmota.github.io/install/ using Tasmota (english)
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
23:35:59.211 MQT: stat/tasmota_F9FF67/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,4704],"FLAG":0,"BASE":18}
23:35:59.435 MQT: stat/tasmota_F9FF67/RESULT = {"Module":{"18":"Generic"}}
23:35:59.673 MQT: stat/tasmota_F9FF67/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"417":"PWM2"},"GPIO5":{"416":"PWM1"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"418":"PWM3"},"GPIO13":{"1216":"AM2301"},"GPIO14":{"192":"Switch_n1"},"GPIO15":{"224":"Relay1"},"GPIO16":{"544":"LedLink"},"GPIO17":{"4768":"ADC Light"}}

```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
23:36:41.826 MQT: stat/tasmota_F9FF67/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":36,"Free":475,"Rules":"ON ANALOG#A0div10 DO status 10 ENDON"}}
23:36:42.053 MQT: stat/tasmota_F9FF67/RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
23:36:42.303 MQT: stat/tasmota_F9FF67/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}

```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
23:38:00.582 MQT: stat/tasmota_F9FF67/STATUS = {"Status":{"Module":18,"DeviceName":"Study Multisensor","FriendlyName":["Study Multisensor","Study Multisensor"],"Topic":"tasmota_F9FF67","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":0,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[15,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
23:38:00.595 MQT: stat/tasmota_F9FF67/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:29:20","StartupUTC":"2022-11-18T12:08:40","Sleep":50,"CfgHolder":4617,"BootCount":15,"BCResetTime":"2022-11-18T11:19:36","SaveCount":185,"SaveAddress":"FB000"}}
23:38:00.605 MQT: stat/tasmota_F9FF67/STATUS2 = {"StatusFWR":{"Version":"12.2.0(tasmota)","BuildDateTime":"2022-10-17T08:39:23","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"395/699"}}
23:38:00.614 MQT: stat/tasmota_F9FF67/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["HOME",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000280","00006000","00004001","00000000"]}}
23:38:00.629 MQT: stat/tasmota_F9FF67/STATUS4 = {"StatusMEM":{"ProgramSize":626,"Free":376,"Heap":21,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
23:38:00.638 MQT: stat/tasmota_F9FF67/STATUS5 = {"StatusNET":{"Hostname":"tasmota-F9FF67-8039","IPAddress":"192.168.0.213","Gateway":"192.168.0.1","Subnetmask":"255.255.255.0","DNSServer1":"1.1.1.1","DNSServer2":"1.0.0.1","Mac":"D8:BF:C0:F9:FF:67","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
23:38:00.648 MQT: stat/tasmota_F9FF67/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.0.40","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_F9FF67","MqttUser":"MQTTuser","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
23:38:00.658 MQT: stat/tasmota_F9FF67/STATUS7 = {"StatusTIM":{"UTC":"2022-11-18T12:38:00","Local":"2022-11-18T23:38:00","StartDST":"2022-10-02T02:00:00","EndDST":"2022-04-03T03:00:00","Timezone":99,"Sunrise":"18:04","Sunset":"03:06"}}
23:38:00.677 MQT: stat/tasmota_F9FF67/STATUS10 = {"StatusSNS":{"Time":"2022-11-18T23:38:00","Switch1":"OFF","ANALOG":{"Illuminance":98},"AM2301":{"Temperature":22.5,"Humidity":43.0,"DewPoint":9.3},"TempUnit":"C"}}
23:38:00.687 MQT: stat/tasmota_F9FF67/STATUS11 = {"StatusSTS":{"Time":"2022-11-18T23:38:00","Uptime":"0T00:29:20","UptimeSec":1760,"Heap":20,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","Dimmer":80,"Color":"CC0000","HSBColor":"0,100,80","Channel":[80,0,0],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"HOME","BSSId":"80:2A:A8:C7:06:C9","Channel":1,"Mode":"11n","RSSI":70,"Signal":-65,"LinkCount":1,"Downtime":"0T00:00:03"}}}
23:38:00.839 MQT: tele/tasmota_F9FF67/STATE = {"Time":"2022-11-18T23:38:00","Uptime":"0T00:29:20","UptimeSec":1760,"Heap":24,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":36,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","Dimmer":80,"Color":"CC0000","HSBColor":"0,100,80","Channel":[80,0,0],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"HOME","BSSId":"80:2A:A8:C7:06:C9","Channel":1,"Mode":"11n","RSSI":70,"Signal":-65,"LinkCount":1,"Downtime":"0T00:00:03"}}
23:38:00.844 MQT: stat/tasmota_F9FF67/RESULT = {"POWER2":"OFF"}
23:38:00.846 MQT: stat/tasmota_F9FF67/POWER2 = OFF
```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
23:38:59.674 MQT: stat/tasmota_F9FF67/RESULT = {"WebLog":4}
23:38:59.823 CFG: Saved to flash at FA, Count 186, Bytes 4096
23:39:00.440 DHT: Pin13 read 01AE0109B9
23:39:04.036 WIF: Checking connection...
23:39:04.443 DHT: Pin13 read 01AE0109B9
23:39:08.453 DHT: Pin13 read 01AE0109B9
23:39:12.459 DHT: Pin13 read 01AE0109B9
23:39:16.419 DHT: Pin13 read 01AE0109B9
23:39:20.423 DHT: Pin13 read 01AE0109B9
23:39:24.021 WIF: Checking connection...
23:39:24.430 DHT: Pin13 read 01AE0109B9
23:39:28.442 DHT: Pin13 read 01AE0109B9
23:39:32.448 DHT: Pin13 read 01AE0109B9
23:39:36.452 DHT: Pin13 read 01AF0109BA
23:39:40.461 DHT: Pin13 read 01AE0108B8
23:39:42.410 WIF: Sending Gratuitous ARP
23:39:44.030 WIF: Checking connection...
23:39:44.420 DHT: Pin13 read 01AD0109B8
23:39:48.429 DHT: Pin13 read 01AE0109B9

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Place cardboard box over the top of the photoresistor, lux drops to 0 in web interface. Leave there for up to 5 minutes waiting for next MQTT telemetry message. Remove box, lux rises to around 300-400 (depending on ambient light level).  Leave there for up to 5 minutes waiting for next MQTT telemetry message. 

### EXPECTED BEHAVIOUR
With the rule in place, I expected to see weblog output that a MQTT telemetry message is sent when the lux level changes significantly. I also would have expected to see it come through in an MQTT client subscribed to the correct topic.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._
Not aoolicable

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment, marked as answer]
With simple analog input, this works as expected, but AFAIK, A0Div10 does not apply when using the analog pin configs parsing the value to match some sensor type. Thus, this behavior is as designed.

**Comment by aheath70** [comment]
Thanks for the suggested answer, and for caring enough to respond - it's appreciated.

You are 100% correct. I changed the A0 definition from ADC Light to ADC Input, and after the restart, the trigger started firing. 

I can work with this, but in an ideal world, just for reporting purposes, it would be good to have the rule be able to be fired when it is defined as A0 Light, or any of the other A0 options. Do you know if there is a technical reason why this isn't the case?

**Reply by sfromis** [reply]
I don't know the history of why it was done the way it is, but a general principle has been to aim for what's "enough" rather than "fill up" with less important polish to add lesser used features which could be nice to have for some users. Tasmota was created for the ESP8266 chip with very limited resources, so a lot of energy has gone into making things as compact as feasible, while still having features for a lot of things.

Here's an example of rules to handle sensor delta change detection and action:
https://tasmota.github.io/docs/Rules/#transmit-sensor-value-only-when-a-delta-is-reached
